### PR TITLE
[feat] implement `record_stream` when using CUDA streams during group offloading

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

With group offloading, we shipped the possibility to overlap computation with communication (data transfers). This PR shows (thanks to the [work](https://gist.github.com/gau-nernst/9408e13c32d3c6e7025d92cce6cba140) of @gau-nernst) that when using CUDA streams, the speed can further be improved with `record_stream`. 

| Record Stream | Memory  | Time   |
|--------------|--------|--------|
| False        | 1.6688 | 32.521 |
| True         | 1.6688 | 31.825 |
 

<details><summary>Code</summary>

```py
from diffusers import DiffusionPipeline
import torch.utils.benchmark as benchmark
from diffusers.hooks import apply_group_offloading
import torch 
import argparse
import json


def benchmark_fn(f, *args, **kwargs):
    t0 = benchmark.Timer(
        stmt="f(*args, **kwargs)",
        globals={"args": args, "kwargs": kwargs, "f": f},
        num_threads=torch.get_num_threads(),
    )
    return f"{(t0.blocked_autorange().mean):.3f}"


def run_inference(pipe, prompt):
    _ = pipe(
        prompt=prompt,
        num_images_per_prompt=1,
        width=576,
        height=1024,
        num_inference_steps=30,
        guidance_scale=3.5,
        max_sequence_length=512
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--record_stream", action="store_true")
    args = parser.parse_args()


    model_id = "black-forest-labs/FLUX.1-dev"
    dtype = torch.bfloat16
    pipe = DiffusionPipeline.from_pretrained(model_id, torch_dtype=dtype)

    apply_group_offloading(
        pipe.transformer,
        offload_type="leaf_level",
        offload_device=torch.device("cpu"),
        onload_device=torch.device("cuda"),
        use_stream=True,
        record_stream=args.record_stream
    )
    apply_group_offloading(
        pipe.text_encoder, 
        offload_device=torch.device("cpu"),
        onload_device=torch.device("cuda"),
        offload_type="leaf_level",
        use_stream=True,
        record_stream=args.record_stream
    )
    apply_group_offloading(
        pipe.text_encoder_2, 
        offload_device=torch.device("cpu"),
        onload_device=torch.device("cuda"),
        offload_type="leaf_level",
        use_stream=True,
        record_stream=args.record_stream
    )
    apply_group_offloading(
        pipe.vae, 
        offload_device=torch.device("cpu"),
        onload_device=torch.device("cuda"),
        offload_type="leaf_level",
        use_stream=True,
        record_stream=args.record_stream
    )

    pipe.set_progress_bar_config(disable=True)

    prompt="A cat wearing sunglasses and working as a lifeguard at pool."
    for _ in range(2):
        image = pipe(
            prompt=prompt,
            num_images_per_prompt=1,
            width=576,
            height=1024,
            num_inference_steps=30,
            guidance_scale=3.5,
            max_sequence_length=512,
            generator=torch.manual_seed(0)
        ).images[0]
    time  = benchmark_fn(run_inference, pipe, prompt)
    inference_memory = torch.cuda.max_memory_allocated() / (1024 ** 3)

    image = pipe(
        prompt=prompt,
        num_images_per_prompt=1,
        width=576,
        height=1024,
        num_inference_steps=30,
        guidance_scale=3.5,
        max_sequence_length=512,
        generator=torch.manual_seed(0)
    ).images[0]

    filename = f"record_stream@{args.record_stream}"
    image.save(f"{filename}.png")
    with open(f"{filename}.json", "w") as f:
        json.dump(
            {"record_stream": args.record_stream, "memory": f"{inference_memory:.4f}", "time": time}, f
        )
```

</details>

Doesn't impact the quality.

Please note that this is just a PoC PR wherein I wanted to show the benefits of using record streams (it was easier with a PoC PR for me than opening a feature request). If this is good, I won't mind buttoning it up further myself for making it merge-ready or if the other maintainers directly commit to the PR branch for doing so. If the improvements are not good enough, I am happy to close the PR without asking anything.

### CODE REVIEW

**Review by a-r-r-o-w — COMMENTED** [review]
Thanks, I think we could consider adding. I did think about it during initial addition of group offloading, but like you reported, the time difference seemed very low, and I attributed it to randomness because it wasn't a ~20%-ish difference (as shown in gau-nerst's gist) and because I didn't spend time actually looking at the stream profiles.

Could you also run the benchmarks again with a warmup and the changes from #11094 and #11097?

**Comment by sayakpaul** [comment]
Even though the `benchmark` module accounts for the warmup ([reference one](https://pytorch.org/docs/stable/benchmark_utils.html), [reference two](https://pytorch.org/tutorials/recipes/recipes/benchmark.html)), I added it to my snippet (updated snippet is included in the OP). 

I merged the PR branches into this PR branch and obtained the following numbers on the DGX:

```bash
{"record_stream": false, "memory": "1.3514", "time": "32.065"}
{"record_stream": true, "memory": "1.3514", "time": "30.901"}
```

On audace, I got:

```bash
{"record_stream": false, "memory": "1.3514", "time": "430.398"}
{"record_stream": true, "memory": "1.3514", "time": "429.695"}
```

File: src/diffusers/hooks/group_offloading.py [hunk 1]
```diff
@@ -68,36 +69,61 @@ def __init__(
...
             # Wait for previous Host->Device transfer to complete
             self.stream.synchronize()
 
         with context:
             for group_module in self.modules:
-                group_module.to(self.onload_device, non_blocking=self.non_blocking)
+                for param in group_module.parameters():
+                    param.data = param.data.to(self.onload_device, non_blocking=self.non_blocking)
+                    if self.record_stream:
+                        param.data.record_stream(current_stream)
```
Comment (a-r-r-o-w) [inline src/diffusers/hooks/group_offloading.py:96]: current_stream should be self.stream here I think. We are telling pytorch that the param.data and buffer.data here is owned by the non-default stream. Currently, we're telling it that it is owned by the default stream, which seems incorrect to me

Sorry for the back and forth but I think we will have to run the benchmark once more with the change 😅 

Apart from this, everything else looks good. We can button up the docs and merge after @DN6 gives a look. Let's make sure to mention that this may use more memory in comparison to `record_stream=False` in certain cases and link to the [torch::Tensor::record_stream](https://pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html) docs

File: src/diffusers/hooks/group_offloading.py [hunk 2]
```diff
@@ -68,36 +69,61 @@ def __init__(
...
+        self.record_stream = record_stream
         self.cpu_param_dict = cpu_param_dict
         self.onload_self = onload_self
 
         if self.stream is not None and self.cpu_param_dict is None:
-            raise ValueError("cpu_param_dict must be provided when using stream for data transfer.")
+            raise ValueError("`cpu_param_dict` must be provided when using stream for data transfer.")
+
+        if self.record_stream and not self.stream:
+            raise ValueError("`record_stream` cannot be True when `stream` is None.")
```
Comment (a-r-r-o-w) [inline src/diffusers/hooks/group_offloading.py:80]:
```suggestion
        if self.stream is not None and self.record_stream:
            raise ValueError("`record_stream` cannot be True when `stream` is None.")
```

(nit): making the check look the same as the one above it for cpu_param_dict

File: src/diffusers/hooks/group_offloading.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/hooks/group_offloading.py:96]: @a-r-r-o-w no problem. Better to be rigorous and double-check everything. 

Just ran the benchmark after merging `main` into this branch (on DGX on a single 80GB A100):

```bash
{"record_stream": false, "memory": "1.3514", "time": "32.792"}
{"record_stream": true, "memory": "1.3514", "time": "30.944"}
```

Feel free to run it yourself if you want. 

> Let's make sure to mention that this may use more memory in comparison to record_stream=False in certain cases and link to the [torch::Tensor::record_stream](https://pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html) docs

Absolutely. I will mention in the docstrings. Is there any other place you wanted me to mention it?

**Comment by sayakpaul** [comment]
Will let @DN6 review and will propagate the Doc changes after that.

**Comment by sayakpaul** [comment]
@a-r-r-o-w @DN6 I have pulled in the latest improvements done to group offloading in this PR and resolved the conflicts. Have also run the benchmarks:

```bash
{"record_stream": true, "memory": "1.3514", "time": "31.047"}
{"record_stream": false, "memory": "1.3514", "time": "32.213"}
```

**Comment by DN6** [comment]
PR looks fine to merge once docstrings are added.

**Review by DN6 — APPROVED** [review]
